proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1195 @@
+        bits = values.assert_next(i32)
+        for i, field in enumerate(chunk):
+          result[field] = bits & (1 << i) != 0
+      yield(result as flags<$fields>)
+    }
+
+    record<$fields> => {
+      yield(record_start<$fields>)
+      for $field in $fields:
+        yield from lift(direction, src, values)
+      yield(record_end<$fields>)
+    }
+
+    variant<$cases> => {
+      discrim = values.assert_next(i32)
+      assert_or_trap(discrim >= 0 && discrim < $cases.len())
+      yield(variant_start<$cases>(discrim))
+      match $cases[discrim] {
+        payload => {
+          # Consume all other values that this variant might consume
+          payload_types = flatten_cases(direction, $cases)
+          payload_values = []
+          for ty in payload_types:
+            payload_values.push(values.assert_next(ty))
+
+          # If there a payload type for this case then we lift it here.
+          # "downcast" the values to the right types for this payload and then
+          # do the conversion
+          if payload:
+            my_types = flatten(direction, payload)
+            payload_values.truncate(my_types.len())
+            for i, ty in enumerate(my_types):
+              payload_values[i] = downcast(payload_values[i], ty)
+            yield from lift(direction, src, payload, payload_values)
+
+          yield(variant_end<$cases>)
+        }
+      }
+    }
+
+    # Lifting a handle means that the module has given us an index into its
+    # tables of resources. We consult the global TABLES helper to ensure that
+    # the `val` given is indeed a valid resource for `src` of type `$resource`
+    handle<$resource> => {
+      val = values.assert_next(i32)
+      yield(TABLES.get($resource, val, src) as handle<$resource>)
+
+      # If we're lifting for an exported function then we're lifting the results
+      # of the export after the call. Similar to lists the handle index
+      # reference is owned by the callee (which just returned) but is a sort of
+      # "temporary" allocation that the callee doesn't need. As a result, like
+      # lists, we automatically remove the index from the source's index space.
+      if direction == "export":
+        TABLES.remove($resource, val, src)
+    }
+
+    # Buffers are only lifted from imports, which means that this only applies
+    # to the parameters of imported functions. This means that the
+    # representation we're dealing with is three i32 values.
+    #
+    # The first i32 is either 0 meaning that this is a new local buffer within
+    # `src`, or it's a nonzero value meaning that it's a handle to a buffer
+    # defined in some other module.
+    #
+    # If the first handle value is 0 then the offset parameter is in bytes from
+    # the start of the memory of `src`. The `len` parameter is the length, in
+    # units of $ty, of the buffer. The pointer must be well aligned.
+    #
+    # If the first handle value is not zero then TABLES.get_buffer will validate
```

Comment:
Not having followed the origin of this design (apologies), why are these 2 types of specifying a buffer a single IT? Seems to me the common case of local memory should be kept as simple as possible.

Also out of interest, since these are fixed length, if you have a variable amount of data to receive, would you typically use some kind of "call again if the number of bytes read is equal to the size of the buffer" approach, or would you use `list<u8>` instead? What if the intended data is utf-8?

Reply:
Nah no worries, these are good questions! I'm not entirely sure what you mean by 2 types as a single interface type, but I'll try to explain a bit more about the background of these types. If I miss the point of your question though let me know!

The `push-buffer` was added for the WASI `fd_read` sycall effectively, and `pull-buffer` for `fd_write`. The purpose is to expose the ability of short reads/writes in the API. The other main motivation for these being separate types is for virtualization and efficiency purposes. For example if `read` were to return a `list<u8>` then that would imply allocating a list. Similarly if `write` took a `list<u8>` then those who virtualized it would receive the entire input buffer copied into their own memory. With `{push,pull}-buffer` these are solved since everything is a "lightweight pointer" where the callee decides how to read from it and how to fill it in.

Reading over this again, the `push/pull` buffer types are not intended to be used interchangeably. They're intended to be separate types. I handled them in a similar branch here because their handling is similar, but I realize now that I probably need some tweaks here and there for the push/pull distinction (e.g. if you pass an index for a push buffer as a value of type pull-buffer then that should trap). I'll fix that locally and push up a clarification.

Reply:
I didn't mean push vs pull, but local memory vs handle (as encoded by the first i32, described by the comments this refers to). That seems to imply conditionals and verification on the receiving end, which at least naively sounds undesirable for just filling a buffer with bytes.

Reply:
Ah apologies!

The goal here was mainly around forwarding where if you expose a method that takes a `push-buffer` then you're called with an `i32` handle. If you want to forward that to something you import then all you have to pass is an `i32`. Forwarding requires them to be the same type, and so the discrimant byte was added to allow you to pass received handles directly to other modules if desired (to avoid intermediate copies and such).

This does require lifting/lowering to check the i32, however, and that would indeed be a cost baked in, but just for the canonical ABI I think. Once fully fleshed out adapter instructions and such are here then a custom ABI could be used where it's always a local buffer (just a pointer/length) and that's wrapped up and passed along.

Reply:
I guess in WASI it is anticipated that passing a buffer thru multiple boundaries is common enough that this "escape hatch" is necessary? One could also argue that for the purpose of being "canonical" you want the simplest, common use case, and that the cost of an extra copy is what will be solved by full adapters.

But anyway, that is probably a discussion for elsewhere. Thanks for the clarification!